### PR TITLE
Adding ChangeStar model

### PR DESCRIPTION
**Pull request by Z-Zheng**
- Adding ChangeStar network architecture from ["Change is Everywhere: Single-Temporal Supervised Object Change Detection in Remote Sensing Imagery", Zheng et al. (ICCV 2021)](https://arxiv.org/abs/2108.07002) and [Offical Code](https://github.com/Z-Zheng/ChangeStar)
- Fixing a typo in docstring of `torchgeo.models.FarSeg`

### CODE REVIEW

File: tests/models/test_changestar.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    @pytest.mark.parametrize("backbone", BACKBONE)
+    def test_changestar_farseg_backbone(self, backbone: str) -> None:
+        if backbone == "anynet":
+            match = "unknown backbone: anynet."
+            with pytest.raises(ValueError, match=match):
+                ChangeStarFarSeg(
+                    classes=4, backbone="anynet", backbone_pretrained=False
+                )
+        else:
+            ChangeStarFarSeg(classes=4, backbone=backbone, backbone_pretrained=False)
```
Comment (adamjstewart) [inline tests/models/test_changestar.py:83]: I think it would be simpler to split this into 2 tests, one that tests all valid args, and one that tests an invalid arg.

File: tests/models/test_changestar.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import itertools
+
+import pytest
+import torch
+from torch.nn.modules import (
```
Comment (adamjstewart) [inline tests/models/test_changestar.py:8]: Shouldn't these be imported directly from `torch.nn`? Also, if you `import torch.nn as nn` instead, it would greatly reduce the number of lines of code.

File: torchgeo/models/changestar.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+from .farseg import FarSeg
+
+Identity.__module__ = "nn.Identity"
+
+
+class ChangeMixin(Module):
+    """This module enables any segmentation model to detect binary change.
+
+    'Change is Everywhere: Single-Temporal Supervised Object Change Detection',
+    Zheng et al. (2020)
```
Comment (adamjstewart) [inline torchgeo/models/changestar.py:29]: I would remove this. We already link to the paper itself, so the title of the paper is redundant.

File: torchgeo/models/changestar.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        in_channels: int = 128 * 2,
+        inner_channels: int = 16,
+        num_convs: int = 4,
+        scale_factor: float = 4.0,
+    ):
+        """Initializes a new ChangeMixin module.
+
+        Args:
+            in_channels: sum of channels of bitemporal feature maps
+                (default=256)
```
Comment (adamjstewart) [inline torchgeo/models/changestar.py:47]: These defaults can be removed, they are already added to the documentation by Sphinx

File: torchgeo/models/changestar.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+            inner_channels: number of channels of inner feature maps
+                (default=16)
+            num_convs: number of convolution blocks
+                (default=4)
+            scale_factor: number of upsampling  factor
+                (default=4.0)
+        """
+        super(ChangeMixin, self).__init__()  # type: ignore[no-untyped-call]
```
Comment (adamjstewart) [inline torchgeo/models/changestar.py:55]:
```suggestion
        super().__init__()  # type: ignore[no-untyped-call]
```
This is the Python 3 syntax for `super()`, Python 2 is long since dead.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: tests/models/test_changestar.py (hunk 2 above)
Comment (Z-Zheng) [reply tests/models/test_changestar.py:8]: Directly using `torch.nn as nn` will raise mypy issues.

Comment (adamjstewart) [reply tests/models/test_changestar.py:8]: We do it in other files, see `torchgeo/models/fcn.py`

**Comment by Z-Zheng** [comment]
It seems to involve other test cases.
How to avoid irrelevant test cases? @adamjstewart 

FAILED tests/datasets/test_chesapeake.py::TestChesapeake13::test_plot - _tkin...
==== 1 failed, 457 passed, 22 skipped, 12 deselected in 177.67s (0:02:57) =====

**Comment by calebrob6** [comment]
Hey @Z-Zheng, I'll look into this now

File: torchgeo/models/changestar.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    UpsamplingBilinear2d,
+)
+
+from .farseg import FarSeg
+
+Identity.__module__ = "nn.Identity"
+
+
+class ChangeMixin(Module):
+    """This module enables any segmentation model to detect binary change.
```
Comment (isaaccorley) [inline torchgeo/models/changestar.py:27]: Could you clarify this more? Seems like it's just feeding images through conv layers. Particularly the 'enables any segmentation model' part.

File: torchgeo/models/changestar.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        c12, c21 = torch.split(
+            c1221, batch_size, dim=0
+        )  # type: ignore[no-untyped-call]
+        return [c12, c21]
+
+
+class ChangeStar(Module):
+    """The base class of the network architecture of ChangeStar.
+
+    ChangeStar is composited of an any segmentation model and a ChangeMixin module.
```
Comment (isaaccorley) [inline torchgeo/models/changestar.py:100]: ~~composited~~ composed

File: torchgeo/models/changestar.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    This model is mainly used for binary/multi-class change detection under bitemporal
```
Comment (isaaccorley) [inline torchgeo/models/changestar.py:101]: Is it actually multi-class? Seems like in `ChangeMixin` you are hardcoding it to output a binary mask. Could we possibly add a `output_channels` arg to `ChangeMixin` to allow for multi-class change detection?

File: torchgeo/models/changestar.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        super().__init__()  # type: ignore[no-untyped-call]
+        self.dense_feature_extractor = dense_feature_extractor
+        self.seg_classifier = seg_classifier
+        self.changemixin = changemixin
+
+        if inference_mode not in ["t1t2", "t2t1", "mean"]:
+            raise ValueError(f"Unknown inference_mode: {inference_mode}")
+        self.inference_mode = inference_mode
+
+    def forward(self, x: Tensor) -> Dict[str, Any]:
```
Comment (isaaccorley) [inline torchgeo/models/changestar.py:140]: Why `Dict[str, Any]` and not `Dict[str, Tensor]`?

File: torchgeo/models/changestar.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        bi_feature = self.dense_feature_extractor(x)
+        # semantic segmentation
+        bi_seg_logit = self.seg_classifier(bi_feature)
+        bi_seg_logit = rearrange(bi_seg_logit, "(b t) c h w -> b t c h w", t=t)
+
+        bi_feature = rearrange(bi_feature, "(b t) c h w -> b t c h w", t=t)
+        # change detection
+        c12, c21 = self.changemixin(bi_feature)
+
+        results: Dict[str, Any] = {}
```
Comment (isaaccorley) [inline torchgeo/models/changestar.py:167]: Same as above

File: torchgeo/models/changestar.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        self.inference_mode = inference_mode
+
+    def forward(self, x: Tensor) -> Dict[str, Any]:
+        """Forward pass of the model.
+
+        Args:
+            x: a bitemporal input tensor of shape [B, T, C, H, W]
+
+        Returns:
+            a directory containing
```
Comment (isaaccorley) [inline torchgeo/models/changestar.py:147]: ~~directory~~ dictionary

File: torchgeo/models/changestar.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+                (default='resnet50')
+            classes: number of output segmentation classes
+                (default=1 for binary segmentation)
+            backbone_pretrained: whether to use pretrained weight for backbone
+                (default=True)
+        """
+        model = FarSeg(
+            backbone=backbone, classes=classes, backbone_pretrained=backbone_pretrained
+        )
+        seg_classifier: Any = model.decoder.classifier
```
Comment (isaaccorley) [inline torchgeo/models/changestar.py:217]: Wny `Any` and not `Module`?

File: torchgeo/models/changestar.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+        model.decoder.classifier = (
+            Identity()  # type: ignore[no-untyped-call, assignment]
+        )
+
+        super(ChangeStarFarSeg, self).__init__(
```
Comment (isaaccorley) [inline torchgeo/models/changestar.py:222]:
```suggestion
        super().__init__(
```

File: torchgeo/models/changestar.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+                    {
+                        "change_prob": torch.stack([c12, c21], dim=0)
+                        .sigmoid_()
+                        .mean(dim=0)
+                    }
+                )
+            return results
+
+        results.update({"bi_seg_logit": bi_seg_logit, "bi_change_logit": [c12, c21]})
+        return results
```
Comment (isaaccorley) [inline torchgeo/models/changestar.py:185]:
```suggestion
        if not self.training:
            results.update({"bi_seg_logit": bi_seg_logit})
            if self.inference_mode == "t1t2":
                results.update({"change_prob": c12.sigmoid()})
            elif self.inference_mode == "t2t1":
                results.update({"change_prob": c21.sigmoid()})
            elif self.inference_mode == "mean":
                results.update(
                    {
                        "change_prob": torch.stack([c12, c21], dim=0)
                        .sigmoid_()
                        .mean(dim=0)
                    }
                )
        else:
           results.update({"bi_seg_logit": bi_seg_logit, "bi_change_logit": [c12, c21]})
        return results
```

**Review by isaaccorley — COMMENTED** [review]
Some minor suggestions. I appreciate the use of einops!

File: torchgeo/models/changestar.py (hunk 12 above)
Comment (Z-Zheng) [reply torchgeo/models/changestar.py:217]: just for avoiding mypy issues.
That's Ok if `Module` can work.

File: torchgeo/models/changestar.py (hunk 8 above)
Comment (Z-Zheng) [reply torchgeo/models/changestar.py:101]: Oh, Sorry. Only binary here.

Comment (Z-Zheng) [reply torchgeo/models/changestar.py:101]: ChangeMixin is used to detect binary change, while the segmentation model provides semantic labels.
The semantic changes can be inferred by a binary change mask and two semantic masks.
Therefore, this network architecture can support binary and multi-class change detection.

File: torchgeo/models/changestar.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+from torch import Tensor
+from torch.nn.modules import (
+    BatchNorm2d,
+    Conv2d,
+    Identity,
+    Module,
+    ReLU,
+    Sequential,
+    UpsamplingBilinear2d,
+)
```
Comment (adamjstewart) [inline torchgeo/models/changestar.py:19]: I would use `import torch.nn as nn` and `nn.BatchNorm2d`, etc. for these too just like with the tests.

Comment (Z-Zheng) [reply torchgeo/models/changestar.py:19]: such as nn.modules.BatchNorm2d?

Comment (Z-Zheng) [reply torchgeo/models/changestar.py:19]: such as `nn.modules.BatchNorm2d`?

Comment (Z-Zheng) [reply torchgeo/models/changestar.py:19]: Seems like directly using nn.BatchNorm2d will raise mypy issues.

Comment (adamjstewart) [reply torchgeo/models/changestar.py]: I think that's a bug in PyTorch's type hints (or maybe in mypy) so you can either use `nn.modules.BatchNorm2d` or use `nn.BatchNorm2d` and ignore the mypy error with a comment.

Comment (Z-Zheng) [reply torchgeo/models/changestar.py]: torchvision has a [mypy configuration file](https://github.com/pytorch/pytorch/blob/master/mypy.ini).
Maybe, adding this file can solve these mypy issues fundamentally.

Comment (adamjstewart) [reply torchgeo/models/changestar.py]: We also have some configuration in https://github.com/microsoft/torchgeo/blob/main/pyproject.toml#L44

I like the idea of disabling some of these warnings in configuration. I probably won't get to this in the next couple of weeks, so use whatever you want for now and I can change it later.

Comment (isaaccorley) [reply torchgeo/models/changestar.py]: I had similar problems with type hints when adding other change detection models. There are workarounds that you can reference here https://github.com/microsoft/torchgeo/blob/main/torchgeo/models/fccd.py.

Comment (Z-Zheng) [reply torchgeo/models/changestar.py]: Thanks, that's very helpful.

File: torchgeo/models/changestar.py (hunk 8 above)
Comment (isaaccorley) [reply torchgeo/models/changestar.py:99]: Ok that makes sense. It might be useful to add that somewhere in the docstring.

Comment (Z-Zheng) [reply torchgeo/models/changestar.py:99]: Added.
Thanks for the suggestion.

File: torchgeo/models/changestar.py [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+                ``'t2t1'``: concatenate bitemporal features in the order of t2->t1;
+                ``'mean'``: the weighted mean of the output of ``'t1t2'`` and ``'t1t2'``
+        """
+        super().__init__()  # type: ignore[no-untyped-call]
+        self.dense_feature_extractor = dense_feature_extractor
+        self.seg_classifier = seg_classifier
+        self.changemixin = changemixin
+
+        if inference_mode not in ["t1t2", "t2t1", "mean"]:
+            raise ValueError(f"Unknown inference_mode: {inference_mode}")
```
Comment (adamjstewart) [inline torchgeo/models/changestar.py:139]: This line is missing test coverage

File: torchgeo/models/changestar.py [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+
+        results: Dict[str, Any] = {}
+        if not self.training:
+            results.update({"bi_seg_logit": bi_seg_logit})
+            if self.inference_mode == "t1t2":
+                results.update({"change_prob": c12.sigmoid()})
+            elif self.inference_mode == "t2t1":
+                results.update({"change_prob": c21.sigmoid()})
+            elif self.inference_mode == "mean":
+                results.update(
```
Comment (adamjstewart) [inline torchgeo/models/changestar.py:177]: These lines are missing test coverage

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: torchgeo/models/changestar.py (hunk 17 above)
Comment (Z-Zheng) [reply torchgeo/models/changestar.py:177]: Added.

File: torchgeo/models/changestar.py (hunk 16 above)
Comment (Z-Zheng) [reply torchgeo/models/changestar.py:139]: Added

**Review by adamjstewart — APPROVED** [review]
Looks good from my end, I think @isaaccorley has some comments that still need to be addressed.

**Comment by isaaccorley** [comment]
Looks like the forward in ChangeStar still needs to be updated to return Dict[str, Tensor] instead of Dict[str, Any]. After that I'm good with this.

File: torchgeo/models/changestar.py [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+                    {
+                        "change_prob": torch.stack([c12, c21], dim=0)
+                        .sigmoid_()
+                        .mean(dim=0)
+                    }
+                )
+        else:
+            results.update(
+                {"bi_seg_logit": bi_seg_logit, "bi_change_logit": [c12, c21]}
```
Comment (calebrob6) [inline torchgeo/models/changestar.py:186]: This is the only thing preventing the method signature from being `Dict[str, Tensor]`, can `c12` and `c21` be combined, or split into different dict keys?

Comment (isaaccorley) [reply torchgeo/models/changestar.py:186]: We could either torch.stack([c12, c21]) the tensors together or just make the type: Dict[str, Union[Tensor, List[Tensor]]. I think either is ok but whatever we choose should be explicitly stated in the docstring so users don't have to go to the source to understand what's happening.

Comment (Z-Zheng) [reply torchgeo/models/changestar.py]: Thanks, revised.

**Review by calebrob6 — APPROVED** [review]

File: torchgeo/models/changestar.py (hunk 18 above)
Comment (isaaccorley) [reply torchgeo/models/changestar.py]: Lgtm

**Comment by calebrob6** [comment]
Awesome! Thanks again @Z-Zheng
